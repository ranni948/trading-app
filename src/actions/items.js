import uuid from 'uuid';
import database, { storage } from '../firebase/firebase';

//ADD_EXPENSE
export const addItem = (item) => ({
    type: 'ADD_ITEM',
    item
});

export const startAddItem = (itemData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            title = '',
            description = '', 
            price = 0, 
            category = '', 
            location = '',
            images = []
        } = itemData;
        const item = { title, description, price, category, location, images };

        return database.ref(`users/${uid}/items`).push(item).then((ref) => {      

            dispatch(addItem({
                id: ref.key,
                userId: uid,
                ...item
            }));
        });
    };
};

//REMOVE_ITEM
export const removeItem = ({ id } = {}) => ({
    type: 'REMOVE_ITEM',
    id
});

export const startRemoveItem = ({ id } = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        database.ref(`users/${uid}/items/${id}`).once('value').then((snapshot) => {
            return snapshot.val();
        }).then((result) => {
            //delete item from DB first so that any reference to the images is gone, if something
            //goes wrong with deletinf the images it won;t matter since there is no reference to them anymore
            database.ref(`users/${uid}/items/${id}`).remove().then(() => {
                dispatch(removeItem({ id }));
            }).then(() => {
                result.images.map((image, index) => {
                    storage.ref(`images/${image.filename}`).delete();
                });
            });
    
        }); 
    }
}

//EDIT_ITEM
export const editItem = (id, updates) => ({
    type: 'EDIT_ITEM',
    id,
    updates
});

export const startEditItem = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/items/${id}`).update(updates).then(() => {
            dispatch(editItem(id, updates));
        });
    }
}

// SET_ITEMS
export const setItems = (items) => ({
    type: 'SET_ITEMS',
    items
});

export const startSetItems = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users`).once('value').then((snapshot) => {
            const items = [];

            snapshot.forEach((child) => {
                child.forEach((child1) => {
                    child1.forEach((childSnapshot) => {
                        items.push({
                            id: childSnapshot.key,
                            userId: child.key,
                            ...childSnapshot.val()
                        });
                    })
                })
            });
            dispatch(setItems(items));
        });
    };
};