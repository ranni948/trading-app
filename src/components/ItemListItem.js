import React from 'react';
import { NavLink } from 'react-router-dom';
import numeral from 'numeral';

const ItemListItem = ({item}) => (
    <NavLink className="list-items list-item" to={`/view/${item.id}`}>
        <img className="list-item-image" src={item.images[0].url}/>
        <span className="list-item__price">{numeral(item.price / 100).format('$0,0.00')}</span>
        <div>{item.title}</div>
    </NavLink>
);

export default ItemListItem;