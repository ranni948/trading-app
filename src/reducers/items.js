//Items Reducer
const itemsReducerDefaultState = []

export default (state = itemsReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_ITEM':
        console.log('Action items', action.item);
        console.log('State', state);
            return [
                ...state,
                action.item
            ]
        case 'SET_ITEMS':
            return action.items
        // case 'REMOVE_EXPENSE':
        //     return state.filter(({ id }) => id !== action.id)
        // case 'EDIT_EXPENSE':
        //     return state.map((expense) => {
        //         if (expense.id === action.id) {
        //             return {
        //                 ...expense,
        //                 ...action.updates
        //             }
        //         } else {
        //             return expense;
        //         };
        //     });
        default:
            return state
    }
};
