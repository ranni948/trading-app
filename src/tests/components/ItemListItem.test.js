import React from 'react';
import { shallow } from 'enzyme';
import ItemListItem from '../../components/ItemListItem';
import items from '../fixtures/items';

test('should render ItemListItem', () => {
    const wrapper = shallow(<ItemListItem item={items[0]}/>);
    expect(wrapper).toMatchSnapshot();
});