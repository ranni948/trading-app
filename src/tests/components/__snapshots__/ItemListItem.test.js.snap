// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ItemListItem 1`] = `
<NavLink
  activeClassName="active"
  ariaCurrent="true"
  className="list-items list-item"
  to="/view/1"
>
  <img
    className="list-item-image"
    src="111"
  />
  <span
    className="list-item__price"
  >
    $200.00
  </span>
  <div>
    bike
  </div>
</NavLink>
`;
